restli-server/src/main/java/com/linkedin/restli/server/BaseRestLiServer.java

```diff
@@ -233,4 +237,69 @@ private RestLiArgumentBuilder lookupArgumentBuilder(ResourceMethodDescriptor met
     }
     return argumentBuilder;
   }
+
+  /**
+   * Method is used to validate the request's batch size is under
+   * the allowed max batch size which is defined in the server resource.
+   *
+   * @throws RestLiServiceException if request's batch size is larger than the allowed max batch size.
+   */
+  private void validateMaxBatchSize(RestLiRequestData requestData, ResourceMethodDescriptor method,
+      ServerResourceContext context) throws RestLiServiceException
+  {
+    MaxBatchSizeSchema maxBatchSizeAnnotation = method.getMaxBatchSize();
+    if (maxBatchSizeAnnotation == null || !maxBatchSizeAnnotation.isValidate())
+    {
+      return;
+    }
+
+    int requestBatchSize = getRequestBatchSize(requestData, method, context);
+    int maxBatchSize = maxBatchSizeAnnotation.getValue();
+
+    if (requestBatchSize > maxBatchSizeAnnotation.getValue())
+    {
+      throw new RestLiServiceException(HttpStatus.S_400_BAD_REQUEST,
+          String.format("The request batch size: %s is larger than the allowed max batch size: %s for method: %s",
+              requestBatchSize, maxBatchSize, method.getMethodName()));
+    }
+  }
+
+  private int getRequestBatchSize(RestLiRequestData requestData, ResourceMethodDescriptor method,
+      ServerResourceContext context) throws RestLiServiceException
+  {
+    switch (method.getMethodType())
+    {
+      case BATCH_GET:
+      case BATCH_DELETE:
+        return requestData.getBatchKeys().size();
+      case BATCH_UPDATE:
+      case BATCH_PARTIAL_UPDATE:
+        return requestData.getBatchKeyEntityMap().size();
+      case BATCH_CREATE:
+        return requestData.getBatchEntities().size();
+      case BATCH_FINDER:
+        return getBatchFinderCriteriaNumber(method, context);
+      default:
+        return 0;
+    }
+  }
+
+  private int getBatchFinderCriteriaNumber(ResourceMethodDescriptor method,
+      ServerResourceContext context) throws RestLiServiceException
+  {
+    try
+    {
+      List<Parameter<?>> parameterList = method.getParameters();
+      Integer criteriaParamIndex = method.getBatchFinderCriteriaParamIndex();
+      Parameter<?> criteriaParam = parameterList.get(criteriaParamIndex);
+      DataList criteriaList = (DataList) context.getStructuredParameter(criteriaParam.getName());
+      return criteriaList.size();
+    }
+    catch (Exception e)
+    {
+      throw new RestLiServiceException(HttpStatus.S_400_BAD_REQUEST,
+          String.format("The batchFinder method: %s does not provide the correct criteria list: %s .",
+              method.getMethod(), e.getMessage()));
```

Comment:
On second thought, how about moving this validation after we build the arguments? 
Look for `restLiArgumentBuilder.buildArguments` in RestLiMethodInvoker (line 324). 

Validating input params should happen before we check for batch size. Also, the argument builder has special logic to wrap single element into an array (this can happen for protocol 1.0.0).

What do you think?

Reply:
I agree with that validate input params should happen before we check batch size. 
Are you suggesting we move the whole batch size validation into the RestLiMethodInvoker, before rest.li invokes the resource method(line 326)?

Reply:
yes

Reply:
Moved the validation into RestLiMethodInvoker.